proposals/0046-astro-env.md

```diff
@@ -0,0 +1,359 @@
+<!--
+  Note: You are probably looking for `stage-1--discussion-template.md`!
+  This template is reserved for anyone championing an already-approved proposal.
+
+  Community members who would like to propose an idea or feature should begin
+  by creating a GitHub Discussion. See the repo README.md for more info.
+
+  To use this template: create a new, empty file in the repo under `proposals/${ID}.md`.
+  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
+  of the accepted proposal.
+-->
+
+- Start Date: 2024-04-15
+- Reference Issues: https://github.com/withastro/roadmap/issues/837
+- Implementation PR: <!-- leave empty -->
+
+# Summary
+
+Improve DX and security around environment variables in Astro.
+
+# Example
+
+```js
+// astro.config.mjs
+import { defineConfig, envField } from "astro/config"
+
+export default defineConfig({
+  env: {
+    schema: {
+      API_URL: envField.static().private().string(),
+      PUBLIC_FOO: envField.static().public().string({ default: "bar" }),
+      STRIPE_KEY: envField.dynamic().private().string()
```

Comment:
1. How do we support detecting leaks for `.static().private()`? If all static env vars are from `"astro:env/static"`, how to we know if it's being used in client code?
  
   I know you mention this below:

    > **Static variables** ... There's no protection to put in place regarding client-side usage, since `import.meta.env` already handles it.

    Does that mean `import.meta.env.API_URL` will be replaced as `undefined` if it's imported in client code? Otherwise it will error in runtime.


2. Should this be `.string().default("bar")` instead to mimic zod?
3. Like zod, are these fields all required by default unless you use `.optional()` or `.default()`?

Reply:
> Does that mean import.meta.env.API_URL will be replaced as undefined if it's imported in client code? Otherwise it will error in runtime.

Yes I thought we could just return `undefined`. I guess it will trick a few people, so idk if the client can make the server trigger an `AstroError` in development (maybe using `import.meta.hot`?)

> Should this be .string().default("bar") instead to mimic zod?

I don't mind tbh!

> Like zod, are these fields all required by default unless you use .optional() or .default()?

I thought about this and I guess it depends how strict with want to be by default! So we can either provide `.optional()` or `.required()`

Reply:
1. If we want a better error, we'd need to separate out as `"astro:env/static/public"` and `"astro:env/static/private"`, so that we can detect during import time. I'm not sure we can detect on usage time since it's dynamically used.

    Perhaps another idea is to export `getEnv()` like `"astro:env/dynamic"` so we can check if it's referencing a private env var at usage time. I'm not sure otherwise if we leave it as undefined since it can be hard to catch what's wrong. I am curious to hear what others think to though.
---
3. I think making it required by default may be better, so that we can warn if an env var is missing and the user can then decide whether it's intentional.

Reply:
1. I like that approach
2. I'm also in favor of making it required by default